src/features/providers/format.ts

```diff
@@ -0,0 +1,75 @@
+"use strict";
+
+import * as cp from "child_process";
+import * as fs from "fs";
+import * as path from "path";
+import * as vscode from "vscode";
+
+import { Configuration } from "../helpers/configuration";
+import { SQLFluff, SQLFluffCommand } from "../utils/sqlfluff";
+
+export class DocumentFormattingEditProvider {
+  public outputChannel: vscode.OutputChannel;
+  public runner: SQLFluff;
+
+  constructor(outputChannel: vscode.OutputChannel) {
+    this.outputChannel = outputChannel;
+    this.runner = new SQLFluff(this.outputChannel);
+  }
+  async provideDocumentFormattingEdits(
+    document: vscode.TextDocument
+  ): Promise<vscode.TextEdit[]> {
+    const filePath = document.fileName.replace(/\\/g, "/");
+    const rootPath = vscode.workspace.workspaceFolders[0].uri.fsPath.replace(/\\/g, "/");
+    const workingDirectory = Configuration.workingDirectory() ? Configuration.workingDirectory() : rootPath;
+    const textEdits: vscode.TextEdit[] = [];
+    this.outputChannel.appendLine(`Format triggered for ${filePath}`);
+
+    if (Configuration.formatEnabled()) {
+      if (document.isDirty) {
+        await document.save();
+      }
```

Comment:
The reason the `sqlfluff.experimental.format.executeInTerminal` was used here and still marked as experimental was that using directly executing the file would result in the file getting overwritten on disk. This can lead to an error if the user changes the file while the formatting has not completed yet. Do you have any ideas for how to fix that bug?